### PR TITLE
Qlearning2

### PR DESCRIPTION
**Pull request by 4144805**

### CODE REVIEW

File: MeesGame/Gameplay/Level/Level.cs [hunk 1]
```diff
@@ -13,11 +13,19 @@ class Level : GameObjectList, IXmlSerializable
         protected int numColumns = 31;
         protected const int CELL_HEIGHT = 64;
         protected const int CELL_WIDTH = 64;
-        protected TimeSpan timeBetweenActions = TimeSpan.FromMilliseconds(300);
+        public TimeSpan timeBetweenActions = TimeSpan.FromMilliseconds(300);
```
Comment (jensj12) [inline MeesGame/Gameplay/Level/Level.cs]: Create property instead

File: MeesGame/Gameplay/GameObjects/Player.cs [hunk 2]
```diff
@@ -151,6 +152,10 @@ class TimedPlayer : Player
 
         public TimedPlayer(Level level, Point location, int layer = 0, string id = "", int score = 0) : base(level, location, layer, id, score)
         {
+            this.score = score;
+            this.level = level;
+            this.location = location;
+            this.position = tileField.GetAnchorPosition(location);
```
Comment (jensj12) [inline MeesGame/Gameplay/GameObjects/Player.cs]: Code duplication, already done in Player class

**Review by jensj12 — CHANGES_REQUESTED** [review]

File: MeesGame/AI/QLearning.cs [hunk 3]
```diff
@@ -1,85 +1,271 @@
...
+using System.Threading;
+using Microsoft.Xna.Framework.Input;
+using System.Linq;
 
 namespace AI
 {
-    // Tijdelijk type om errors te voorkomen, moet uiteindelijk ergens in de MeesGame namespace worden gedefiniëerd.
-    class ILevelState
+    /// <summary>
+    /// Deze class bevat alle informatie over het level die de qLearning AI nodig heeft om te leren.
```
Comment (maxmeijer42) [inline MeesGame/AI/QLearning.cs:12]: English please

File: MeesGame/AI/QLearning.cs [hunk 4]
```diff
@@ -1,85 +1,271 @@
...
+        private AILevelState currentState;
+
+        /// <summary>
+        /// learningRate is a number that determines how important it is to try out new ways of completing the maze. 
+        /// If it is very high, it will choose a random action more often. If it is very low, it will often stick to the known best action.
+        /// 
         /// discountFactor is a number that determines how much more important earlier rewards are. 
         /// The lower discountFactor is, the more important it is to the Q function to get rewards as soon as possible.
         /// </summary>
         private double learningRate, discountFactor;
```
Comment (maxmeijer42) [inline MeesGame/AI/QLearning.cs:88]: If you split these variables onto multiple lines you can split the documentation between them

File: MeesGame/AI/QLearning.cs [hunk 5]
```diff
@@ -1,85 +1,271 @@
...
+            else if (inputHelper.IsKeyDown(Keys.G))
+            {
+                //doe het voor het echie, learning rate op 1 zetten zoat hij neit gekke dingen gaat doen
+                learningRate = 1;
+
+                level.timeBetweenActions = TimeSpan.FromMilliseconds(300);
+            }
+        }
+
+        public int numberOfKeys()
```
Comment (maxmeijer42) [inline MeesGame/AI/QLearning.cs]: Method names should always start with a verb. Either make this a property or change the name. Also, the naming convention is to use capital letters..

File: MeesGame/AI/QLearning.cs [hunk 6]
```diff
@@ -1,85 +1,271 @@
...
+            //zorg dat bestActions geupdatet wordt
+            //zorg dat de currentKeyList wordt geupdatet
         }
 
         /// <summary>
-        /// AIPlayGame receives ILevelStates, determines which action it wants to perform and then performs this action.
+        /// AIDecideAction receives AILevelStates, determines which action it wants to perform and then performs this action.
         /// </summary>
-        public void AIPlayGame()
+        //waarom performt deze methode de action al als er een methode AITrainingPerformAction is? Omdat dit níét voor de trainingmodus is maar voor de echte modus.
```
Comment (maxmeijer42) [inline MeesGame/AI/QLearning.cs]: Discussies niet houden in comments in de code

**Comment by jensj12** [comment]
PR contains changes from threaded-AI branch

**Pull request closed without merging**